Fix: don't wrap arrays in `test.each` template strings

yogmel opened the pull request:
<!-- Please provide a brief summary of your changes: -->
Fixes #7653

This adds a verification in the "ArrayPattern" in `printer-estree.js` and updates `shouldBreak` variable if a Jest Each Template Literal is identified.

-----------

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

yogmel commented:
Hi @thorn0!

This is not my best piece of code and I'm looking forward to improving it.

As you mentioned, I have to test all parent nodes of the current AST Node. I performed a for loop for that, but I'm not sure what to use as a limiting factor.

I have sent this PR as I thought it might be easier to review it with real code.

This works with the cases presented by the reported and also doesn't affect current tests. I didn't find any test cases with `test.each` template strings, so I didn't update any existing snapshot.

thorn0 commented:
Hi there! Using the same indices for `path.stack` and `path.getNode` isn't the right thing to do because `stack` contains not only nodes but also property names. (I recommend to run Prettier in a debugger to explore the structure of `path`.)

Anyways, it's better to view `path.stack` as a private property of `FastPath` and use only the methods. Examples of loops similar to what we need can be found here:

https://github.com/prettier/prettier/blob/765e3b40251abbf7f636580dfcb99f69da3dd42d/src/language-js/needs-parens.js#L214-L227

and here:

https://github.com/prettier/prettier/blob/765e3b40251abbf7f636580dfcb99f69da3dd42d/src/language-js/printer-estree.js#L277-L290

> This works with the cases presented by the reported

Good! Please add those cases to the tests (in `tests/js/test-declarations`). As you can see, the CI has failed because your change isn't covered by tests ("codecov/patch — 66.66% of diff hit (target 80.00%)").

And just for the sake of accuracy, note that the nodes we work with in this case have the type `ArrayExpression` (array literals), not `ArrayPattern` (part of the destructuring syntax).

yogmel commented:
Hi @thorn0,

Thanks for the information! Those were pretty useful.

I've changed the loop and added the test cases.

Please let me know if it needs more changes!

thorn0 commented on src/language-js/printer-estree.js:
```diff
@@ -1558,6 +1558,19 @@ function printPathNoParens(path, options, print, args) {
+
+            if (n.type === "ArrayExpression") {
+              let i = 0;
+              let node = path.getNode(i);
```
@yogmel You shouldn't have moved this loop inside the other loop that iterates through the elements of the array literal. Please move it back.

thorn0 commented on tests/js/test-declarations/jest-each.js:
```diff
@@ -60,3 +60,18 @@ test.each([
   }
 );
+
+test.each`
```
Please extract these tests to a separate file. When test files get long, their snapshots are difficult to read.

fisker commented on src/language-js/printer-estree.js:
```diff
@@ -1558,6 +1558,19 @@ function printPathNoParens(path, options, print, args) {
+                if (parent && isJestEachTemplateLiteral(node, parent)) {
+                  return false;
+                }
+                node = parent;
```
This loop back to the root, can we break the loop when meet an expression or something safe?

yogmel replied:
@fisker by root, do you mean moving this logic outside `printPathNoParens()` function?

yogmel commented:
Hi everyone, sorry I am late.

I have extracted the test to another file, but I can't make `yarn test` pass for some reason.

```
Summary of all failing tests
 FAIL  tests/js/test-declarations/jsfmt.spec.js (6.307 s)
  ● template-literal.js › encountered a declaration exception

    Value false is not a valid document

      20 |     !(typeof val === "string" || (val != null && typeof val.type === "string"))
      21 |   ) {
    > 22 |     throw new Error(
         |           ^
      23 |       "Value " + JSON.stringify(val) + " is not a valid document"
      24 |     );
      25 |   }

      at assertDoc (src/document/doc-builders.js:22:11)
          at Array.forEach (<anonymous>)
      at concat (src/document/doc-builders.js:34:11)
      at Object.genericPrint [as print] (src/language-js/printer-estree.js:227:10)
      at callPluginPrintFunction (src/main/ast-to-doc.js:127:18)
      at comments.printComments (src/main/ast-to-doc.js:64:16)
      at Object.printComments (src/main/comments.js:502:19)
      at printGenerically (src/main/ast-to-doc.js:62:22)
      at FastPath.map (src/common/fast-path.js:124:21)
      at printPathNoParens (src/language-js/printer-estree.js:2451:30)
```

thorn0 commented:
The test is failing for a natural reason: something isn't right with the code being tested.

thorn0 commented on src/language-js/printer-estree.js:
```diff
@@ -1576,6 +1577,18 @@ function printPathNoParens(path, options, print, args) {
+        if (n.type === "ArrayExpression") {
+          let i = 0;
+          let node = path.getNode(i);
+          while (node) {
```
```suggestion
          while (node && shouldBreak) {
```

thorn0 commented:
@yogmel Looks good. 👍 Please add a changelog entry (to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`).

yogmel commented:
Thanks for your help and patience, @thorn0
I have submitted the last commit with the changelog regarding this PR.

thorn0 commented on src/language-js/printer-estree.js:
```diff
@@ -1576,6 +1577,18 @@ function printPathNoParens(path, options, print, args) {
 
+        if (n.type === "ArrayExpression") {
+          let i = 0;
+          let node = path.getNode(i);
```
```suggestion
          let node = n;
```

fisker commented:
I still have a concern about performance, I'll have a test.

fisker commented:
I run format over `prettier@2.0.5` files.

The `isJestEachTemplateLiteral` function is called 2061 times, less than I thought it would be.

@thorn0 Do you think it worse to add a cache? With cache reduce to 1334 times.



<details>
<summary>The cache logic I use to test</summary>

```js
const cache = new WeakMap();
function isInsideJestEachTemplateLiteral(path) {
  const parents = new Set();

  let i = 0;
  let node = path.getValue();
  while (node) {
    if (cache.has(node)) {
      const result = cache.get(node);
      for (const node of parents) {
        cache.set(node, result);
      }
      return result;
    }
    parents.add(node);
    const parent = path.getParentNode(i++);
    if (parent && isJestEachTemplateLiteral(node, parent)) {
      for (const node of parents) {
        cache.set(node, true);
      }
      return true;
    }
    node = parent;
  }

  for (const node of parents) {
    cache.set(node, false);
  }
  return false;
}
```
</details>

thorn0 commented:
@fisker Agree, it should make sense to optimize it.

As we're already inside a tree-traversing recursion, it'd be more efficient to add a temporary flag property to `options` when printing `expressions` here:

https://github.com/prettier/prettier/blob/7659a867ad77b493f6c7f10a4421063c03a2a8ef/src/language-js/printer-estree.js#L2358-L2367

and then replace our check with checking that property. What do you think?

fisker commented:
That's a better idea!

yogmel commented:
Hi everyone,
Is this going to be done in this PR? I can do it, but I might need some (more) guidance.
Thanks!

thorn0 commented:
@yogmel Let's do it in this PR. Guidance is not a problem. I'll be glad to help.

yogmel commented:
Thanks @thorn0 :) I will have more time to have a look at it again this weekend. I really appreciate your mentorship.

yogmel commented:
From what I understood, I have to pass a new property to `options`, so that when it passes the `case "ArrayPattern"` operations I can use that to verify, instead of doing the while loop.

The problem I face is that, although the "TemplateLiteral" type is on the outer branch of the object tree, the `case "TemplateLiteral"` routine runs **after** the `case "ArrayPattern"`.

![image](https://user-images.githubusercontent.com/3901809/84222468-0994ce80-aaae-11ea-9507-7e382b99cdfb.png)
(_the test case I'm using is [this one](https://deploy-preview-8354--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuE8DOMB0cCGYAWABgDpQ4AEAPuQEbn0MPVikAkwAjAL6P3XsBtYOQw4ATjCTkArABpyCACZSO08lwC6PfsAEdZAJlkBmWQBZZcgGyyA7LIAcWtpy46hImOMkz5SlWqa2uSCxKhwGGHyYRhiAJZQAOZR5GEAZhBiKWEADmJwMDBxcFkgzlDsxjy87sKiElJyClDK5KrqWlQhuuXsZsGCdV4NbX4tUsbqskPeAWOtBgAMU53u5YQAFGFwAB44ALY5ADZw5OgwKRvAOPI08mBcAJTkALwAfOTAT6QgsiAQOSK0DQyFA4jEEAA7gAFcQIEEoHBHSE4ACeIL+NDEeAA1gUAMo5PAJRLIGBiACucD+cH2NDgikUDIAMjgkhScIk4AAxTL7HCFEnIEA4CkwCC-ED4GD7I4AdXwcXQRLAcHx8KVcQAbkrUcKwGgMSAEmgSjBodjEvzkGkkaa-gArNA7ABC2LAeJg+IOcGZCTgNrt1JATp2+JJJwAihSIPBA0d7SAiWJTWJhTQcPSjpK8gkYHK4ooYPhkA5Fn88hBTXLsTlhXkIiUtQG-gBHGPwC0AhEitAAWigcAZDMl+XbcXyFs51qQtoTwdN+ziZMpC4jcGjsYDs6Dfy8NALRZLSCMqGxcSOJIAwhB9jOQBFpJKKaaACqZhFzxNaqkASRaCBemA8SAgAgi0+IwKiJzxqaXBcEAA)_)

By the time `printJestEachTemplateLiteral()` is called, the Array Expression is already formatted.

Sorry if I misunderstood the task. I forgot how daunting it could be to read a whole new codebase.

thorn0 commented:
You didn't misunderstand anything. The expressions in the placeholders of the template literal are printed (converted from AST to Doc) by this call:
```js
let expressions = path.map(print, "expressions"); 
```
Now it indeed happens before the check that detects `test.each`. We need to change the order of how things are called. The processes of printing `test.each` and other template literals don't necessarily have to share the same `expressions` variable. The printing of the expressions for `test.each` can be moved to `printJestEachTemplateLiteral` and `let expressions = ...` in `case "TemplateLiteral"` can end up after the `test.each` check.

yogmel commented:
Great, thanks for your help! I've sent another update.
This solution looks much better to me. The only part I was not able to solve optimally was this [one](https://github.com/prettier/prettier/pull/8354/commits/46ffe5701512440fdd39539996dc18be3a28f089#diff-3abca41914b4b214a2507e33cc54c28bR3886-R3888).

I thought about passing a parameter to the `print` callback in  `const expressions = path.map(print, "expressions")`, but `printGenerically` does not take the `options` parameter.

thorn0 commented:
Yes, passing a flag via `options` is not exactly elegant, but still it's a good optimization, so let's keep it this way for now. `printGenerically` accepts `args`, but it's passed only one recursion level deep. One day we might implement something like `deepArgs` for passing such things down the tree.

thorn0 commented on src/language-js/printer-estree.js:
```diff
@@ -3873,13 +3876,16 @@ function printMethodInternal(path, options, print) {
 }
 
-function printJestEachTemplateLiteral(node, expressions, options) {
+function printJestEachTemplateLiteral(node, options, path, print) {
```
Let's reduce the number of parameters. We don't need `node` if we have `path`.

thorn0 commented on src/language-js/printer-estree.js:
```diff
@@ -3873,13 +3876,16 @@ function printMethodInternal(path, options, print) {
+  options.isJestTemplateLiteral = false;
   const headerNames = node.quasis[0].value.raw.trim().split(/\s*\|\s*/);
   if (
     headerNames.length > 1 ||
```
Please move printing the expressions inside this `if`. Print results are cached, so if this condition doesn't hold true , the other `print` call for the expressions (in `case "TemplateLiteral"`) will get the same results, which will be wrong.

yogmel commented:
Ok, thanks again @thorn0! Please let me know if it looks ok.

thorn0 commented on src/language-js/printer-estree.js:
```diff
@@ -3873,18 +3875,22 @@ function printMethodInternal(path, options, print) {
 }
 
-function printJestEachTemplateLiteral(node, expressions, options) {
+function printJestEachTemplateLiteral(options, path, print) {
```
Please change the order to `path, options, print` to be consistent with similar `print*` functions.

yogmel commented:
Thanks for the help @thorn0 @fisker !
